File: matiasmasso/Mat.NET/DTO/DTOTpvLog.cs
using MatHelperStd;
using System;
using System.Collections.Generic;

namespace DTO
{
    public class DTOTpvLog : DTOBaseGuid
    {
        public string Ds_Date { get; set; }
        public string Ds_Hour { get; set; }
        public string Ds_Amount { get; set; }
        public string Ds_Currency { get; set; }
        public string Ds_Order { get; set; }
        public string Ds_MerchantCode { get; set; }
        public string Ds_Terminal { get; set; }
        public string Ds_Signature { get; set; }
        public string Ds_Response { get; set; }
        public string Ds_MerchantData { get; set; }
        public string Ds_ProductDescription { get; set; }
        public string Ds_SecurePayment { get; set; }
        public string Ds_TransactionType { get; set; }
        public string Ds_Card_Country { get; set; }
        public string Ds_AuthorisationCode { get; set; }
        public string Ds_ConsumerLanguage { get; set; }
        public string Ds_Card_Type { get; set; }

        // response:
        public string Ds_MerchantParameters { get; set; }
        public string Ds_SignatureReceived { get; set; }
        public string ErrDsc { get; set; }


        public DTOTpvRequest.Modes Mode { get; set; }
        public DTOBaseGuid Request { get; set; }
        public DTOCca Result { get; set; }
        public DTOUser User { get; set; }
        public DTOContact Titular { get; set; }
        public DateTime FchCreated { get; set; }
        public bool SignatureValidated { get; set; }
        public bool ProcessedSuccessfully { get; set; }

        public string Exceptions { get; set; }


        public DTOTpvLog() : base()
        {
        }

        public DTOTpvLog(Guid oGuid) : base(oGuid)
        {
        }

        public static DTOLang Lang(DTOTpvLog oTpvLog)
        {
            var retval = DTOLang.Factory(DTOLang.Ids.ESP);
            if (TextHelper.VbIsNumeric(oTpvLog.Ds_ConsumerLanguage))
            {
                switch (System.Convert.ToInt32(oTpvLog.Ds_ConsumerLanguage))
                {
                    case 2:
                        {
                            retval = DTOLang.Factory(DTOLang.Ids.ENG);
                            break;
                        }

                    case 3:
                        {
                            retval = DTOLang.Factory(DTOLang.Ids.CAT);
                            break;
                        }
                }
            }
            return retval;
        }

        public static DTOAmt Amt(DTOTpvLog oTpvLog)
        {
            decimal dcAmt = 0;
            string sAmt = oTpvLog.Ds_Amount;
            if (TextHelper.VbIsNumeric(sAmt))
                dcAmt = System.Convert.ToDecimal(sAmt) / 100;
            DTOAmt retVal = DTOAmt.Factory(dcAmt);
            return retVal;
        }

        public static DTOPgcPlan.Ctas CreditCtaCod(DTOContact oTitular)
        {
            var retval = DTOPgcPlan.Ctas.TransferenciesDesconegudes;
            if (oTitular != null)
                retval = DTOPgcPlan.Ctas.Clients_Anticips;
            return retval;
        }

        public static DTOPgcPlan.Ctas DebitCtaCod()
        {
            return DTOPgcPlan.Ctas.VisasCobradas;
        }

        public static bool isValidResponse(DTOTpvLog oLog, List<Exception> exs)
        {
            bool retval = false;
            if (oLog.Ds_Response == "0000")
                retval = true;
            else
                exs.Add(new Exception(string.Format("Codi de error Redsys {0}: {1}", oLog.Ds_Response, oLog.ErrDsc)));

            if (!oLog.SignatureValidated)
            {
                exs.Add(new Exception("Les signatures no coincideixen"));
                retval = false;
            }

            if (!TextHelper.VbIsNumeric(oLog.Ds_AuthorisationCode))
            {
                exs.Add(new Exception("Operació no autoritzada"));
                retval = false;
            }
            return retval;
        }

        public static DTOMailMessage MailMessage(DTOTpvLog oTpvLog)
        {
            var sRecipient = DTOMailMessage.wellknownAddress(DTOMailMessage.wellknownRecipients.Info);
            var sSubject = string.Format("Avís de Tpv: {0}", oTpvLog.Result.Concept);
            var retval = DTOMailMessage.Factory(sRecipient, sSubject);
            return retval;
        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOTpvLog.vb
Public Class DTOTpvLog
    Inherits DTOBaseGuid

    Property Ds_Date As String
    Property Ds_Hour As String
    Property Ds_Amount As String
    Property Ds_Currency As String
    Property Ds_Order As String
    Property Ds_MerchantCode As String
    Property Ds_Terminal As String
    Property Ds_Signature As String
    Property Ds_Response As String
    Property Ds_MerchantData As String
    Property Ds_ProductDescription As String
    Property Ds_SecurePayment As String
    Property Ds_TransactionType As String
    Property Ds_Card_Country As String
    Property Ds_AuthorisationCode As String
    Property Ds_ConsumerLanguage As String
    Property Ds_Card_Type As String

    'response:
    Property Ds_MerchantParameters As String
    Property Ds_SignatureReceived As String
    Property ErrDsc As String


    Property Mode As DTOTpvRequest.Modes
    Property Request As DTOBaseGuid
    Property Result As DTOCca
    Property User As DTOUser
    Property Titular As DTOContact
    Property FchCreated As DateTime
    Property SignatureValidated As Boolean
    Property ProcessedSuccessfully As Boolean

    Property Exceptions As String


    Public Sub New()
        MyBase.New()
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
    End Sub

    Shared Function Lang(oTpvLog As DTOTpvLog) As DTOLang
        Dim retval = DTOLang.Factory(DTOLang.Ids.ESP)
        If TextHelper.VbIsNumeric(oTpvLog.Ds_ConsumerLanguage) Then
            Select Case CInt(oTpvLog.Ds_ConsumerLanguage)
                Case 2
                    retval = DTOLang.Factory(DTOLang.Ids.ENG)
                Case 3
                    retval = DTOLang.Factory(DTOLang.Ids.CAT)
            End Select
        End If
        Return retval
    End Function

    Shared Function Amt(oTpvLog As DTOTpvLog) As DTOAmt
        Dim DcAmt As Decimal
        Dim sAmt As String = oTpvLog.Ds_Amount
        If TextHelper.VbIsNumeric(sAmt) Then
            DcAmt = CDec(sAmt) / 100
        End If
        Dim retVal As DTOAmt = DTOAmt.Factory(DcAmt)
        Return retVal
    End Function

    Shared Function CreditCtaCod(oTitular As DTOContact) As DTOPgcPlan.Ctas
        Dim retval = DTOPgcPlan.Ctas.TransferenciesDesconegudes
        If oTitular IsNot Nothing Then
            retval = DTOPgcPlan.Ctas.Clients_Anticips
        End If
        Return retval
    End Function

    Shared Function DebitCtaCod() As DTOPgcPlan.Ctas
        Return DTOPgcPlan.Ctas.VisasCobradas
    End Function

    Shared Function isValidResponse(oLog As DTOTpvLog, exs As List(Of Exception)) As Boolean
        Dim retval As Boolean
        If oLog.Ds_Response = "0000" Then
            retval = True
        Else
            exs.Add(New Exception(String.Format("Codi de error Redsys {0}: {1}", oLog.Ds_Response, oLog.ErrDsc)))
        End If

        If Not oLog.SignatureValidated Then
            exs.Add(New Exception("Les signatures no coincideixen"))
            retval = False
        End If

        If Not TextHelper.VbIsNumeric(oLog.Ds_AuthorisationCode) Then
            exs.Add(New Exception("Operació no autoritzada"))
            retval = False
        End If
        Return retval
    End Function

    Shared Function MailMessage(oTpvLog As DTOTpvLog) As DTOMailMessage
        Dim sRecipient = DTOMailMessage.wellknownAddress(DTOMailMessage.wellknownRecipients.Info)
        Dim sSubject = String.Format("Avís de Tpv: {0}", oTpvLog.Result.concept)
        Dim retval = DTOMailMessage.Factory(sRecipient, sSubject)
        Return retval
    End Function
End Class
